_posts/2017-06-09-upgrade-to-rails-3.markdown

```diff
@@ -0,0 +1,157 @@
+
+## 5. Config files
+Rails 3 introduces the concept of an Application object. An application object holds all the specific application configurations and it's similar to the current config/environment.rb from Rails 2.3. The application object is defined in config/application.rb. You should move there most of the configuration that you had in config/environment.rb.
+
+In terms of routes, there are a couple of changes that you need to apply to your routes.rb file. For example:
+
+```
+# Rails 2.3 way:
+ActionController::Routing::Routes.draw do |map|
+  map.resources :products
+end
+
+# Rails 3.0 way:
+AppName::Application.routes do
+  resources :products
+end
+```
+You can go to [this article](https://blog.engineyard.com/2010/the-lowdown-on-routes-in-rails-3) to see in depth about this topic.
+
+## 6. Gems
+[Bundler](https://bundler.io/) is the default way to manage Gem dependencies in Rails 3 applications. You will need to add a [Gemfile](https://bundler.io/v1.15/gemfile_man.html) in the root of your app, define all you gems there, and then get rid of the config.gem.
+
+Remember that if you installed the plugin mentioned in step 3, you can run `rake rails:upgrade:gems`. This task will extract your config.gem calls and generate code that you can put into your Gemfile.
+
+## 7. Deprecations
+There are a bunch of deprecations that happen during this version:
+
+#### Active Record
+- The method to define a named scope is now called `scope` instead of `named_scope`.
+
+- In scope methods, you no longer pass the conditions as a hash:
+
+```
+# Before:
+named_scope :active, :conditions => ["active = ?", true]
+
+# Now:
+scope :active, where("active = ?", true)
+```
+
+- `save(false)` is deprecated, so you should use `save(:validate => false)`.
+
+- I18n error messages for Active Record should be changed from `:en.activerecord.errors.template` to `:en.errors.template`.
+
+- `model.errors.on` is deprecated in favor of `model.errors[]`
+
+- There is a new syntax for presence validations:
+
+```
+# Before:
+validates_presence_of :email
+
+# Now:
+validates :email, presence: true
+```
+
+- `ActiveRecord::Base.colorize_logging` and `config.active_record.colorize_logging` are deprecated in favor of `Rails::LogSubscriber.colorize_logging` or `config.colorize_logging`.
+
+### Action Mailer
+- `:charset`, `:content_type`, `:mime_version`, `:implicit_parts_order` are all deprecated in favor of `ActionMailer.default :key => value` style declarations.
+- Mailer dynamic `create_method_name` and `deliver_method_name` are deprecated, just call `method_name` which now returns a `Mail::Message` object.
+
+```
+# Before:
+message = UserMailer.create_welcome_email(user)
```

Comment:
@lubc This is not consistent with the paragraph explanation before this example. I believe it would be clearer if the before example showed this: 

```
UserMailer.deliver_create_welcome_email(user)
```